src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -20,110 +18,128 @@ import com.intellij.openapi.vfs.VirtualFileManager
 import com.intellij.openapi.vfs.newvfs.BulkFileListener
 import com.intellij.openapi.vfs.newvfs.events.VFileEvent
 import com.intellij.util.Alarm
+import com.intellij.util.messages.Topic
 import org.jetbrains.annotations.TestOnly
 import org.rust.cargo.project.CargoProjectDescription
-import org.rust.cargo.project.CargoProjectDescriptionData
 import org.rust.cargo.project.settings.rustSettings
 import org.rust.cargo.project.settings.toolchain
 import org.rust.cargo.project.workspace.CargoProjectWorkspace
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener.UpdateResult
 import org.rust.cargo.toolchain.RustToolchain
 import org.rust.cargo.util.cargoLibraryName
 import org.rust.cargo.util.cargoProjectRoot
+import org.rust.cargo.util.enqueue
 import org.rust.cargo.util.updateLibrary
-import kotlin.properties.Delegates
+import org.rust.lang.utils.lock
+import java.util.concurrent.locks.Lock
+import java.util.concurrent.locks.ReentrantLock
 
-private val LOG = Logger.getInstance(CargoProjectWorkspaceImpl::class.java);
 
-@State(
-    name = "CargoMetadata",
-    storages = arrayOf(Storage(file = StoragePathMacros.MODULE_FILE))
-)
-class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorkspace, PersistentStateComponent<CargoProjectState>, BulkFileListener {
+/**
+ * [CargoProjectWorkspace] component implementation
+ */
+class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorkspace, ModuleComponent {
 
-    private val DELAY_MILLIS = 1000
+    private val DELAY = 1000 /* milliseconds */
+
+    private val LOG = Logger.getInstance(CargoProjectWorkspaceImpl::class.java)
 
     /**
      * Alarm used to coalesce consecutive update requests.
-     * It uses EDT thread, but the tasks are really tiny and
-     * only spawn background update.
+     * It uses pooled thread.
      */
     private val alarm = Alarm()
 
-    private var cargoProjectState: CargoProjectState by Delegates.observable(CargoProjectState()) {
-        prop, old, new ->
-            projectDescription = new.projectData?.let { CargoProjectDescription.deserialize(it) }
-    }
+    /**
+     * Lock to guard reads/updates to [cached]
+     */
+    private var lock: Lock = ReentrantLock()
```

Comment:
Looks are hard. I don't feel confident using looks (unless in Rust :) ), so I think it is a good idea to avoid them. 

And I think that [AtomicReferece](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/AtomicReference.html) is a much simpler synchronization primitive which can be used here. 

Why is it sufficient?  Tasks are executed in serialized order and, moreover, `onSuccess` handler is invoked on `EDT` thread. And `onSuccess` is (or should be) the only place where we modify `cached`. So we have a single writer / multiple readers situation, so we don't need to enforce mutual exclusion and only need to care for liveness. So `AtomicReference` fits the bill, and, what is crusually important guarantees the absence of dead locks by construction. Another interesting property is that you can't forget to lock the lock if you use atomic.